Testing: Await E2E promise interactions

aduth opened the pull request:
Extracted from #13569 

This pull request seeks to fix a few occurrences within E2E test specs where we use async functions without properly awaiting their resolution. This should help with overall E2E test stability.

Notably:

- `block-hierarchy-navigation.test.js` has been a frequent intermittent failure in builds (failed snapshot around "Third Column"). I have reasonably high confidence the changes here will resolve these.
- I have low-moderate confidence the `change-detection.test.js` could resolve some failed test runs which result in messages along the lines of "the context has already been destroyed"

**Implementation notes:**

I sought these out using a combination of regular expression file searches of increasing breadth within `e2e-tests/specs`, identifying a pattern where a line does not begin with `await` but calls a function which may be async:

```regex
^\s*page\.waitFor
```
```regex
^\s*page\.
```
```regex
^\s*(?!describe|expect|it|beforeEach|before|after|afterEach)\w+\(
```

It would be nice if there were some way to automate this detection, as in almost all cases, a function which is asynchronous or which otherwise returns a promise should be `await`ed before proceeding (exceptions can include occurrences within `Promise.all`, etc).

**Testing instructions:**

Verify E2E tests run:

```
npm run test-e2e
```

aduth committed:
Testing: Await E2E promise interactions

gziolo commented:
> It would be nice if there were some way to automate this detection, as in almost all cases, a function which is asynchronous or which otherwise returns a promise should be awaited before proceeding (exceptions can include occurrences within Promise.all, etc).

https://www.npmjs.com/package/eslint-plugin-jest

I don't see any Jest specific ESLint rules which could help here. It's very tricky.

gziolo commented:
Would https://eslint.org/docs/rules/require-await help here in some ways?

gziolo commented:
I triggered re-run but I still see failure for one of the tests:

https://travis-ci.org/WordPress/gutenberg/jobs/501527708#L1120

```
AIL packages/e2e-tests/specs/manage-reusable-blocks.test.js (33.413s)
   ● Managing reusable blocks › Should import reusable blocks
     waiting for XPath "div[@class="post_title"][contains(text(), "Greeting")]" failed: timeout 30000ms exceeded
```

aduth commented:
>It seems to be an actual issue as this test was updated.

I suspect then it's an issue which probably existed prior to the test here, but only became obvious now that we await the resolution of the `waitForXPath`

gziolo commented:
> I suspect then it's an issue which probably existed prior to the test here, but only became obvious now that we await the resolution of the waitForXPath

Should we remove this check or is there an easy way to fix it?

aduth committed:
Testing: Disable animations on every admin screen

aduth committed:
Testing: Refactor reusable block management to compare before/after entry count

aduth commented:
> Should we remove this check or is there an easy way to fix it?

I'm not confident enough with XPath to know what was wrong with the selector, or how it could be improved. In testing locally, the import worked correctly, but the XPath never matched. My hunch is that it was previously testing an element which was non-visible on the screen, where [`page.waitForXPath`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitforxpathxpath-options) has some options to this effect.

In any case, I refactored the test in 380d89a to something which seems a bit more durable anyways.

I also had some issues with intermittent failures locally related to animations conflicting with button-clicking. Our "disable animations" previously only applied to the editor screen, not the reusable blocks screen. I changed it in 76c63c0 to disable animations everywhere. This should also help for #13569, where I needed to change this plugin as well (see https://github.com/WordPress/gutenberg/pull/13569#issuecomment-468829853).

gziolo reviewed (approved):
Yes, those improvements make a lot of sense. In particular, using `getNumberOfEntries` seems less fragile than the previous XPath based verification. Let's :shipit: as it should make e2e tests more solid immediately.

aduth commented:
> Would https://eslint.org/docs/rules/require-await help here in some ways?

I don't think it helps for this, but I do think it might be worth including in the default ruleset regardless. I was looking at it the other day, in fact, since there's currently nothing to stop someone from adding `async` to every function in our code. The "When Not To Use It" section from the documentation seems to imply there are some legitimate counter-examples, albeit largely edge cases.

aduth commented:
> I don't see any Jest specific ESLint rules which could help here. It's very tricky.

Yeah, it's not something which would be easy to statically analyze, since it requires knowledge of referenced functions enough (possibly in other files) to know whether they're async.
